Scripts: use default babel if none is found in project

oandregal opened the pull request:
Changes to make Webpack use the `@wordpress/babel-preset-default` babel config if none is found when running `wp-scripts build/start`.

## Test

- make sure you can build Gutenberg for development: `npm run dev` and that it works as expected.
- make sure you can build Gutenberg for production: `npm run build` and that it works as expected.
- external testbed using the default babel config https://github.com/nosolosw/understanding-gutenberg/pull/7

oandregal committed:
Extract webpack utils from build command

oandregal committed:
Do not export hasWebpackConfig

as it is not used anywhere else.

oandregal committed:
Webpack will use @wordpress/babel-preset-default if no babel config found

oandregal committed:
Teach Webpack where to lookup for @wordpress/babel-preset-default

Starts in own @wordpress/scripts node_modules.

gziolo commented on packages/scripts/config/webpack.config.js:
```diff
@@ -58,6 +58,21 @@ const externals = [
+	exclude: /node_modules/,
+	use: {
+		loader: require.resolve( 'babel-loader' ),
+		options: {
```
The only difference here are `options` so maybe it would be possible to consolidate this code in a way where only `options` are provided conditionally.

nosolosw replied:
addressed at f7c1a04

gziolo commented on packages/babel-preset-default/index.js:
```diff
@@ -3,31 +3,31 @@ module.exports = function( api ) {
 	return {
 		presets: [
-			! isTestEnv && [ '@babel/preset-env', {
+			! isTestEnv && [ require.resolve( '@babel/preset-env' ), {
```
I did some research and I found that `babel-loader` has some issues with the way how paths are being resolved and this is one of the workarounds often shared by the community. However, if we go this route we might need to update more places in our codebase. One of them is here:
https://github.com/WordPress/gutenberg/blob/master/bin/packages/get-babel-config.js#L16

Ideally, we don't have to change anything in the preset itself.

nosolosw replied:
I've found a way to make this work for 3rd parties and Gutenberg. TLDR is that this proposal centralizes all babel logic into the `@wordpress/babel-preset-default`.

oandregal committed:
Centralize babel logic in preset

oandregal committed:
Use api.caller instead of env variables

nosolosw commented on bin/packages/get-babel-config.js:
```diff
@@ -1,64 +1,38 @@
+	 * The caller options can only be 'boolean', 'string', or 'number' by design:
+	 * https://github.com/babel/babel/blob/bd0c62dc0c30cf16a4d4ef0ddf21d386f673815c/packages/babel-core/src/config/validation/option-assertions.js#L122
+	 */
+	const callerOpts = { caller: {
```
Before this, I've tried to use an object shape like: 

```
{ caller: { 
    name: 'WP_BUILD'
    presetEnv: {
        // options to change would go here
    },
    pluginTransformRuntime: {
        // options to change would go here
    }
} }
```

This would be more flexible and future-proof. Could even be a basis for bringing extensibility for 3rd parties. We can't do that at the moment [by design](https://github.com/babel/babel/blob/bd0c62dc0c30cf16a4d4ef0ddf21d386f673815c/packages/babel-core/src/config/validation/option-assertions.js#L122) so added a note to keep an eye on that.

nosolosw commented on bin/packages/build.js:
```diff
@@ -164,9 +164,7 @@ function buildScssFile( styleFile ) {
 	const buildDir = BUILD_DIR[ environment ];
 	const destPath = getBuildPath( file, buildDir );
-	const babelOptions = getBabelConfig( environment );
-	babelOptions.sourceMaps = true;
```
Moved this within the `getBabelConfig` function for coherence.

gziolo replied:
It's fine, we just need to ensure it gets synced with https://github.com/WordPress/gutenberg/pull/14409 :)

nosolosw replied:
Rebased from master to get this in.

nosolosw commented on packages/babel-preset-default/index.js:
```diff
@@ -1,36 +1,78 @@
+	// on the caller options, we need to tell the cache
+	// when it needs updating. Otherwise, it won't be
+	// recalculated for different builds.
+	api.cache.using( () => wpBuildOpts.name );
```
The babel cache needs to be updated every time the `wpBuildOpts.name` changes. We need to do this to make sure the preset is updated. The [`bin/packages/build.js` script](https://github.com/WordPress/gutenberg/blob/664781a47820d6c37c0ef8366ac9b9d293e28b13/bin/packages/build.js) currently iterates over each source file and generates both bundles (`main` and `modules`), so this means the cache won't be used.

I like to time `bin/packages/build.js` script to measure if this makes a noticeable difference.

nosolosw replied:
> I like to time `bin/packages/build.js` script to measure if this makes a noticeable difference.

Compared this branch to `master` and both take about the same time to execute `npm run dev` and `npm run build`.

gziolo replied:
I can't see `api.cache` used in any of the Babel presets: https://github.com/babel/babel/search?p=3&q=api.cache&unscoped_q=api.cache

We probably should skip it as well.

nosolosw replied:
We are already using `api.env` in `babel-preset-default`, which updates the `api.cache` automatically for us. [See](https://babeljs.io/docs/en/config-files#apienv):

> This function internally makes use of api.cache mentioned below to ensure that Babel is aware that this build depends on a specific envName.

Once we introduce variable preset (different config options depending on what we want to build) we need to update the cache manually so it returns the proper results. We could perhaps rewrite the current approach to use node env variables and `api.env` but will be updating the cache in the preset anyway.

My reading of the situation is that we're fine here.

nosolosw replied:
So, it looks like the [`api.caller`](https://babeljs.io/docs/en/config-files#apicallercb) is also used by the cache to determine whether to re-compute the preset if it isn't already parsed. We don't need to manually update it, just returning the proper key.

For `npm run build` and `npm run dev`, I've checked that the preset is only parsed three times: for the keys `WP_BUILD_MAIN`, `WP_BUILD_MODULE`, and `undefined`. The first two are parsed when the `bin/packages/build.js` is executed. The other one happens when webpack calls `babel-loader` (which doesn't pass any `caller` option).

Note that `api.test` will also update the cache if the environment is different. This only happens when we run tests. I think we can perhaps simplify this to only use the `api.caller`, but I'd like to do that in a different PR.

oandregal committed:
Refactor checks

gziolo commented on packages/babel-preset-default/index.js:
```diff
@@ -1,36 +1,75 @@
+			};
+		}
+
+		if ( isWPBuild( wpBuildOpts.name ) ) {
```
Can this check be moved to line 26 as suggested?

gziolo commented on packages/babel-preset-default/index.js:
```diff
@@ -1,36 +1,75 @@
+		if ( isTestEnv ) {
+			opts.useBuiltIns = 'usage';
+		} else {
+			opts.modules = false;
```
```suggestion
			opts.modules = isWPBuild( wpBuildOpts.name ) ? wpBuildOpts.module : false;
```

nosolosw replied:
That would change the current logic. We want `modules=commonjs` independently of what `isTestEnv` value is.

gziolo commented on packages/babel-preset-default/index.js:
```diff
@@ -1,36 +1,75 @@
+
+		const opts = {
+			helpers: true,
+			useESModules: false,
```
How about:

```suggestion
			useESModules: wpBuildOpts.name === 'WP_BUILD_MODULE' ? wpBuildOpts.useESModules : false,
```

nosolosw replied:
I prefer to leave it as it is so it's similar to what we have in `getPresetEnv`.

oandregal committed:
Changes from rebase

oandregal committed:
Update how browserslist-config package is looked up

It is only used by babel-preset-default. It is looked up by require
itself so we need to tell Jest not to substitute with anything.

oandregal committed:
Teach babel-preset-default where to lookup for deps

oandregal committed:
Update comment

oandregal committed:
Fix browserslist-config dep

oandregal committed:
Simplify which code changes

oandregal committed:
The babel cache uses the api.caller

So we don't need to update it manually.

gziolo commented on packages/babel-preset-default/index.js:
```diff
@@ -1,36 +1,74 @@
 module.exports = function( api ) {
+	let wpBuildOpts = {};
+	const isWPBuild = ( name ) => {
+		const result = [ 'WP_BUILD_MAIN', 'WP_BUILD_MODULE' ].some(
```
`result` variable can be inlined, the way it was before 😄

nosolosw replied:
Blended it together with other things at e7dc7f1

gziolo commented on test/unit/jest.config.json:
```diff
@@ -1,6 +1,7 @@
 {
 	"rootDir": "../../",
 	"moduleNameMapper": {
+		"@wordpress\\/(browserslist-config)$": "@wordpress/$1",
```
I have PR coming up soon which will redefine `moduleNameMaper` completely to make it bullet-proof, so let's fly with this for now.

oandregal committed:
Be explicit about cache keys

gziolo commented on packages/scripts/config/webpack.config.js:
```diff
@@ -66,6 +66,8 @@ const externals = [
 const isProduction = process.env.NODE_ENV === 'production';
 const mode = isProduction ? 'production' : 'development';
 
+const getBabelLoaderOptions = () => hasBabelConfig() ? {} : { presets: [ require.resolve( '@wordpress/babel-preset-default' ) ] };
```
I figured out we should ensure it doesn't load any random configs when we provide our own. `react-scripts` does it as follows:

https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/config/webpack.config.js#L354-L355

nosolosw replied:
Done at 06f132b

It shouldn't pick anything because we check for the babel config first. Yet, I've tested that `hasBabelConfig` is stricter than babel itself (a file named `.babelrc.js` won't be detected as babel config by `hasBabelConfig` by it will by babel). So: added those checks to be future-proof.

Also added `.babelrc.js` as a valid babel config in dbbc04b.

oandregal committed:
Do not use any config file if we provide the config

oandregal committed:
Add .babelrc.js file to the list of valid babel configs

See https://babeljs.io/docs/en/config-files#file-relative-configuration

oandregal committed:
Delete filed added in rebase

gziolo commented on packages/scripts/utils/config.js:
```diff
@@ -7,6 +7,7 @@ const { hasPackageProp } = require( './package' );
 
 const hasBabelConfig = () =>
 	hasProjectFile( '.babelrc' ) ||
+	hasProjectFile( '.babelrc.js' ) ||
```
Nice catch :)

oandregal committed:
Remove node_modules from packages.

Individual packages shouldn't have node_modules installed in
their local paths in a normal Gutenberg operation. Yet,
if you `cd packages/<packagename> && npm install` things start
going south - like tests not passing.

gziolo reviewed (approved):
I tested it very extensively and it all works perfectly fine. Thi is AMAZING. Outstanding work on making it all work!!!

🥇 💯 🚀
